Groups v2

fancypantalons opened the pull request:
This PR represents a refactoring that culminates in control over whether groups are automatically opened/joined at startup.

First, we refactoring all the tracking so that groups are stored alongside the SignaldAccount in their own list.  Previously group data was attached to the conversation, but since we may not have open conversations for all groups right away, we have to start tracking that information separately.

Second, we automatically add group chats to the buddy list and mark them persistent.  Since we're not automatically opening chats, this is the only way to ensure the user has a way to open a group chat that isn't opened yet.  Marking them persistent ensures that when you close the chat, you don't accidentally leave it unless you really mean it!

Third, since we're in here adding a bunch of capabilities anyway, we add support for renaming group chats.  We do this using the Pidgin "Alias..." function, such that if you alias a chat it's renamed in the Signal protocol and all the associated tracking is updated.  Similarly, if we receive a rename, we automatically alias the chat to rename it.

Then, last but not least, we add the new auto-join-group-chats setting.  When disabled, group chats are only automatically opened when a message is received (this ensures we don't lose any messages; if we didn't open the conversation, we'd have nowhere to put them!)  When enabled, they're opened at startup as per the previous behaviour.

The default is to *not* automatically open chats.  So Bitlbee users will definitely want to enable this setting.

Commit:
Refactoring before some behavioural changes.

Ultimately my goal is to improve group chat auto-join behaviour by
making things a bit more configurable.  In particular, I want the plugin
to support having a record of an existing conversation but it not being
joined yet, so that joins can optionally be done on-demand from the
buddy list.

But first I needed to refactor a bunch of this code (again).

Prior to this commit, information about the group chats--and in
particular the user list for those chats--was being carried around by
Pidgin PurpleConversation instances.  This meant that a
PurpleConversation instance had to be created when we discovered a group
chat.

Obviously that doesn't work if there isn't a 1-to-1 relationship between
PurpleConversation instances and Signal groups.

So this change adds a bunch of additional bookkeeping for groups so that
we track this metadata separately from the Pidgin internals.

The next step will then be to implement the behaviour enhancements.  But
since this rewiring is so significant, I wanted to get it done, first,
before pushing on, just in case I really broke things...

Commit:
Fix likely segfault.

Spotted this while reading through the code.  Obviously if
g_hash_table_remove is freeing the SignaldGroup instance, it's not a
wise move to access the group to pull the id out...

Commit:
Only update conversation user list if it's open.

In the future we may get group membership updates before the
conversation is open, so we need to check for that.

Commit:
Fix a NULL dereference in some circumstances.

When diffing the member list in an UPDATE against our record of the
group, if a group update contains only additions or removals (i.e., what
usually happens), the other list comes back NULL.  We need to check for
that...

Commit:
Add joined group chats to the buddy list.

Subject says it all.  Now when we join a group chat it's automatically
added to the buddy list (and marked as persistent).

Commit:
Remove the chat when we leave it.

This also lays a bunch of groundwork for the final steps for better
controlling chat opening.

Commit:
Add support for renaming Signal groups.

I'm not sure this is the "correct" approach, but now if you Alias a
chat, this renames the Signal chat.  Similarly, if the group is renamed
remotely, we reflect the update by aliasing the chat locally.

Commit:
Control whether chats are opened at startup.

This addresses issue #13 by adding a new setting (auto-join-group-chats)
that controls whether or not group chats are automatically joined/opened
at startup.

If the setting is turned off, group chats are only opened if a message
is received.  I've opted to do this to ensure no messages are lost (you
need the conversation open so you have a place to put the messages as
they arrive!)

If the setting is turned on, group chats are opened at startup (the
previous behaviour), which is preferable for bitlbee.

The default is for this setting to be turned *off*.

demure commented:
This change probably merits some form of documentation of the settings, either in the readme or perhaps the start of a docs/ dir.

fancypantalons commented:
> This change probably merits some form of documentation of the settings, either in the readme or perhaps the start of a docs/ dir.

IMO the latter would be ideal.  There definitely needs to be a setup/install guide for linking, and the settings for external attachment hosting also deserve to be documented.

Commit:
Clean up some old test code.

Commit:
Actually write message after opening convo.

We auto-open group chats when we receive a message, but if we open the
chat, we should actually write the message...

hoehermann commented:
Does this include the changes from commit 5a99200 mentioned in #13 or am I to merge and cherry-pick both?

Commit:
Delete old group chats that we're not a member of.

If the user leaves a group while Pidgin isn't running, we might have an
old group chat kicking around in the buddylist that we need to delete.

fancypantalons commented:
> Does this include the changes from commit [5a99200](https://github.com/hoehermann/libpurple-signald/commit/5a9920046908808f746c750e9408bc40abaadd93) mentioned in #13 or am I to merge and cherry-pick both?

This PR contains everything (including an additional fix I just pushed to deal with orphaned group chats).  So you can just review and pull this and you should get everything!

hoehermann commented:
Nice work! Your code is very clean _and_ it even has comments! :)

demure commented:
Did the new settings end up in any user facing documentation?

hoehermann commented:
No. Not as far as I know. At least I did not write anything. ;)

Moreover, it is unlikely that I will supply a detailed user-documentation for these new features. I am writing a lot of documentation at my day job and I do not want this project to feel like work.

fancypantalons commented:
Hah, I get that.  I have no issues throwing a little bit of doc together if it's helpful.  If @demure if there isn't already an issue filed to create some doc, feel free to file one and I can dig into it at some point!

Of course, if you wanted to volunteer a first draft... ;)

demure commented:
I can take a stab at documenting it from my bitlbee bias perspective. I assume markdown format, in a `/doc` dir is desirable?

hoehermann commented:
Yes, markdown in a doc directory is good.
